### PR TITLE
feat: use cuda api wrappers instead of cuda utils

### PR DESCRIPTION
**Pull request by wep21**
## Description

use [cuda-api-wrappers](https://github.com/eyalroz/cuda-api-wrappers) which is more generalized cuda wrapper than cuda utils.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: common/cuda_api_wrappers/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,37 @@
+cmake_minimum_required(VERSION 3.17)
+project(cuda_api_wrappers)
+
+find_package(CUDAToolkit)
+if(NOT TARGET CUDA::cudart OR NOT TARGET CUDA::cuda_driver)
+  message(WARNING cuda runtime library or cuda dirver library is not found.)
```
Comment (kenji-miyake) [inline common/cuda_api_wrappers/CMakeLists.txt:6]: typo

```suggestion
  message(WARNING cuda runtime library or cuda driver library is not found.)
```

**Comment by wep21** [comment]
@kenji-miyake Does `nvcc` exist in `ghcr.io/autowarefoundation/autoware-universe:humble-latest-cuda`?

**Comment by kenji-miyake** [comment]
@wep21 Yes, it seems so. Why? :thinking: 

```console
❯ docker run --rm -it ghcr.io/autowarefoundation/autoware-universe:humble-latest-cuda
root@281a252c886d:/autoware# nvcc -V
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2022 NVIDIA Corporation
Built on Tue_Mar__8_18:18:20_PST_2022
Cuda compilation tools, release 11.6, V11.6.124
Build cuda_11.6.r11.6/compiler.31057947_0
root@281a252c886d:/autoware#
```

File: common/cuda_api_wrappers/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+find_package(CUDAToolkit)
+if(NOT TARGET CUDA::cudart OR NOT TARGET CUDA::cuda_driver)
+  message(WARNING cuda runtime library or cuda driver library is not found.)
+  return()
+endif()
+
+include(FetchContent)
+fetchcontent_declare(cuda-api-wrappers
+  GIT_REPOSITORY https://github.com/eyalroz/cuda-api-wrappers.git
+  GIT_TAG 831666a0bfd1af0f44f4fa234ee2d983d347fcaa # v0.6.1
```
Comment (yukkysaito) [inline common/cuda_api_wrappers/CMakeLists.txt:13]: I think it would be better to import with vcs, but what do you think?

Comment (wep21) [reply common/cuda_api_wrappers/CMakeLists.txt:13]: I'm fine with moving this vendor package to repos file, but I would not like to increase the number of packages in repos as possible.

**Comment by wep21** [comment]
@kenji-miyake It seems that `CMAKE_CUDA_COMPILER` is not set in [the ci with cuda](https://github.com/autowarefoundation/autoware.universe/actions/runs/4090253939/jobs/7053532714).
```
CMake Error at /__w/autoware.universe/autoware.universe/build/cuda_api_wrappers/_deps/cuda-api-wrappers-src/CMakeLists.txt:19 (PROJECT):
  No CMAKE_CUDA_COMPILER could be found.
```
In my environment, this variable is set to the path of `nvcc`.

File: common/cuda_api_wrappers/CMakeLists.txt (hunk 2 above)
Comment (yukkysaito) [reply common/cuda_api_wrappers/CMakeLists.txt:16]: https://github.com/autowarefoundation/autoware/blob/main/autoware.repos#L28-L69
we currently use vcs to manage external libraries etc., so I think vcs is better than downloading with cmake, what do you think?

Comment (yukkysaito) [reply common/cuda_api_wrappers/CMakeLists.txt:16]: I don't really like the idea of not being able to build without an internet connection. If possible, I would like to finish the process of requiring a internet connection when the environment is built. (Right now, it also is required at build time for DNN weight file downloads.)

File: common/cuda_api_wrappers/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,40 @@
+cmake_minimum_required(VERSION 3.17)
```
Comment (manato) [inline common/cuda_api_wrappers/CMakeLists.txt:1]: @wep21 
Thank you for the suggestion. It looks a cool alternative.
According to [README](https://github.com/eyalroz/cuda-api-wrappers#requirements), it seems to require cmake version 3.18 or later. So how about to set the required version to `cmake_minimum_required` ?
```suggestion
cmake_minimum_required(VERSION 3.18)
```

File: perception/tensorrt_yolox/src/tensorrt_yolox.cpp [hunk 4]
```diff
@@ -71,20 +68,22 @@ TrtYoloX::TrtYoloX(
...
     const auto output_dims = trt_common_->getBindingDimensions(1);
-    input_d_ = cuda_utils::make_unique<float[]>(batch_config[2] * input_size);
+    input_d_ = cuda::memory::device::make_unique<float[]>(batch_config[2] * input_size);
     out_elem_num_ = std::accumulate(
       output_dims.d + 1, output_dims.d + output_dims.nbDims, 1, std::multiplies<int>());
     out_elem_num_per_batch_ = static_cast<int>(out_elem_num_ / batch_config[2]);
-    out_prob_d_ = cuda_utils::make_unique<float[]>(out_elem_num_);
-    out_prob_h_ = cuda_utils::make_unique_host<float[]>(out_elem_num_, cudaHostAllocPortable);
+    out_prob_d_ = cuda::memory::device::make_unique<float[]>(out_elem_num_);
+    out_prob_h_ = cuda::memory::host::make_unique<float[]>(out_elem_num_);
```
Comment (manato) [inline perception/tensorrt_yolox/src/tensorrt_yolox.cpp:76]:
```suggestion
    out_prob_h_ = cuda::memory::host::make_unique<float[]>(out_elem_num_);
    cuda::memory::host::register_(out_prob_h_.get(), out_prob_h_.size(), CU_MEMHOSTREGISTER_PORTABLE);
```
I guess this `register_` call might be needed to make an equivalent code (?) (Sorry for my unfamiliarity with the usage of this wrapper). I'd appreciate your consideration.

**Comment by Mingyu1991** [comment]
@wep21 
This PR is not active for 4 months and it seems not all suggestions are resolved.
Is this PR still valid?

**Comment by wep21** [comment]
@Mingyu1991 Thank you for your ping. I give up merging the PR at this time.

**Pull request closed without merging**